vllm/model_executor/guided_decoding.py

```diff
@@ -0,0 +1,112 @@
+import asyncio
+import concurrent.futures
+from copy import copy
+from enum import Enum
+from functools import lru_cache
+from json import dumps as json_dumps
+from re import escape as regex_escape
+from typing import Union, Tuple
+from types import SimpleNamespace
+from pydantic import BaseModel
+
+from vllm.entrypoints.openai.protocol import CompletionRequest, ChatCompletionRequest
+from vllm.model_executor.guided_logits_processors import JSONLogitsProcessor, RegexLogitsProcessor
+
+
+class GuidedDecodingMode(Enum):
+    JSON = "json"
+    REGEX = "regex"
+    CHOICE = "choice"
+
+
+async def get_guided_decoding_logits_processor(
+        request: Union[CompletionRequest, ChatCompletionRequest],
+        tokenizer
+    ) -> Union[JSONLogitsProcessor, RegexLogitsProcessor]:
+    """
+    Given an OpenAI-compatible request, check for guided decoding parameters
+    and get the necessary logits processor for the given guide.
+    We cache logit processors by (guide, tokenizer), and on cache hit
+    we make a shallow copy to reuse the same underlying FSM.
+    """
+    guide, mode = _get_guide_and_mode(request)
+    if not guide:
+        return None
+    loop = asyncio.get_running_loop()
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=2) as pool:
+        result = await loop.run_in_executor(
+            pool, _get_cached_logits_processor,
+            guide, tokenizer, mode
+        )
+        logits_processor = copy(result)
+        # reset logits processor's internal state
+        logits_processor.init_state()
+        return [logits_processor]
+    
+
+def _get_guide_and_mode(
+        request: Union[CompletionRequest, ChatCompletionRequest]
+    ) -> Tuple[str, GuidedDecodingMode]:
+    # validate guided decoding parameters
+    guide_count = sum([
+        request.guided_json is not None,
+        request.guided_regex is not None,
+        request.guided_choice is not None
+    ])
+    if guide_count == 0:
+        return None, None
+    elif guide_count > 1:
+        raise ValueError(
+            "You can only use one kind of guided decoding "
+            "('guided_json', 'guided_regex' or 'guided_choice')."
+        )
```

Comment:
This validation can be moved to the protocol.py in pydantic model definition. https://docs.pydantic.dev/latest/concepts/validators/#model-validators